Add UI event transactions (#766)
getsentry/develop: src/docs/sdk/performance/ui-event-transactions.mdx

* __Idle-transactions__: The SDK starts an idle timeout for the transaction when starting it.
When the transaction or any of its spans starts a new span, it resets the timeout. The SDK finishes
the transaction when the timeout succeeds.
* __Wait-for-children__: A transaction waits for all its child spans to finish before finishing itself.

Before diving into the specification with all the edge cases, let's look at two simple examples:

1. The user clicks a button that triggers some requests to the backend and stores data in the
local database. The SDK can create a meaningful transaction in that case.
2. The user clicks a button that solely validates some form data and triggers nothing that the
SDK can automatically instrument. The SDK could only create a transaction without any spans, which would
be valueless.

# Specification

Users can change the `idleTimeout` via the SDK config options. The default is 3.0 seconds. 

The specification is written in the [Gherkin syntax](https://cucumber.io/docs/gherkin/reference/).

```Gherkin
Scenario: Starting UI Event transactions
    Given an instrumentable UI event
    Then the SDK starts a UI event transaction 

>     And starts the idle timeout with idleTimeout of the options

Reviewer: ```suggestion
    And schedules to finish with idleTimeout of the options
```
`l`: maybe this is clearer a bit?

Scenario: Wait for children when starting span
    Given a UI event transaction
    When the SDK starts a child span
    Then the SDK cancels the idle timeout
    And waits for the child to finish

> Scenario: Start timeout when the last span finishes

Reviewer: `l`: Maybe we should use `schedule` everywhere instead of `start`?

Reviewer 2: Yeah, a Timer is scheduled.

    Given a UI event transaction
    And the transaction has one or multiple running child spans
    And the transaction is waiting for its children to finish
    When the SDK finishes the last child span
    Then the SDK starts the idle timeout

Scenario: Don't reset timeout when the second last span finishes
    Given a UI event transaction
    And the transaction has two running child spans
    When the SDK finishes the first child span
    Then the SDK doesn't reset the idle timeout
    And waits for the children to finish

Scenario: Discard UI event transactions without child spans
    Given a UI event transaction
    And the transaction has no child spans
    When the idleTimeout times out
    Then the SDK discards the transaction 

Scenario: Set time to last finished child span
    Given a UI event transaction
    And the transaction has one finished child span
    And the transaction has one running child span
    When the running child span finishes

>     Then the SDK finishes the transaction

Reviewer: ```suggestion
    Then the SDK schedules the idle timeout
    And the SDK finishes the transaction after timeout
```
`m`: maybe this will help better understand why we need to trim the end time - because transaction is not finished immediately with the last child span

    And trims the end time of the transaction to the one of the last finished
        child span

Scenario: Same UI element with same event
    Given an ongoing UI event transaction
    When the user triggers the same UI event with the same type for the same
        UI element
    Then the SDK resets the timeout
    And doesn't create a new transaction

> # If your SDK binds auto-genrated transactions to the scope see binding

Reviewer: ```suggestion
# If your SDK binds auto-generated transactions to the scope see binding
```

# to scope
Scenario: Same UI element with different event
    Given an ongoing UI event transaction
    When the user triggers the same UI element with a different event
    Or the user triggers a different UI element

>     Then the SDK finishes the ongoing transaction

Reviewer: `h`: I think it's important to mention here that we also remove it from scope (on Android we do), because calling `finish` doesn't necessarily trigger the immediate finish due to `waitForChildren`

```suggestion
    Then the SDK finishes the ongoing transaction
    And removes the transaction from scope
```

Reviewer: ahh, just noticed it's in the different section, sorry my bad!

    And sets the status to OK
    And waits for the children to finish
    And cancels the timeout
    And starts a new transaction
```

## Binding to scope

On platforms mainly interacting with the static API, such as mobile, it's typical to bind auto-generated
transactions to the scope so that users can access them via the static API. We recommend binding UI event
transactions to the scope on these platforms. The following extra rules apply as UI event transactions
could interfere with other auto-generated transactions.

```Gherkin
Scenario: Same UI element with different event
    Given an ongoing UI event transaction
    When the user triggers the same UI element with a different event
    Or the user triggers a different UI element
    Then the SDK finishes the ongoing transaction
    And sets the status to OK
    And waits for the children to finish
    And cancels the timeout
    And removes the ongoing transaction from the scope
    And starts a new transaction
    And puts the new transaction on the scope

Scenario: UI event triggered but transaction ended

>     Given an auto-generated transaction from a any UI event

Reviewer: ```suggestion
    Given an auto-generated transaction from any UI event
```
is it a typo or my english is not that advanced? 🤣

Author: Typo thanks 😁🙏

>     And the transaction already  is finished

Reviewer: ```suggestion
    And the transaction is already finished
```

    When the user triggers the same UI event
    Then the SDK starts a new UI event transaction

Scenario: Manually created transaction bound to the scope
    Given an ongoing manually created transaction by the user bound to the
        scope
    When the SDK creates a new UI event transaction

>     Then the SDK doesn't add the UI event transaction to the scope 

Reviewer 2: In this case, if the transaction is not in the scope, there will be no spans for it since the user can't access it nor the auto instrumentation, doesn't that mean that the transaction won't have spans and won't be sent?

Author: Basically, yes, we could drop it and not start it. That's an improvement. Good catch, thanks 🙏 . @romtsn, I think we could change that in our SDKs. WDYT?

Reviewer 2: I guess it makes sense if it's not an Idle transaction (for platforms that you know that the loading is finished), but not with the idle transaction concept.

Reviewer: yeah we could do that, I guess that would work for any kind of transaction on the scope - if there's something, we just don't start the UI transaction at all

Reviewer 2: Should we raise a meta issue for that?

Reviewer: https://github.com/getsentry/team-mobile/issues/71

```

### Screen load transactions

This section deals specifically with auto-generated transactions for loading screens. If your SDK has
other types of auto-generated transactions, please update the specification here.

```Gherkin
Scenario: Ongoing UI event transaction
    Given an ongoing UI event transaction
    When the SDK creates a new screen load transaction
    Then the SDK finishes the ongoing UI event transaction
    And removes it from the scope
    And sets the status to canceled
    And waits for its children to finish

Scenario: Ongoing screen load transaction
    Given an ongoing screen load transaction
    When the SDK starts a new UI event transaction
    Then the SDK doesn't bind the new UI event transaction to the scope

> ```

Reviewer: @marandaneto I guess we can also add a navigation transaction here (we have it for compose, and you for flutter). Or is it considered a screen load transaction?

Reviewer 2: I guess it can be both, it's just a concept but they behave in the same way.
`ui.load` when you know when to start/likely when to finish.
`navigation` when it's a router/navigation API.

## Transaction name

The user should be able to identify to which UI element the UI event transaction belongs on which screen
by only looking at the transaction name. Choose whatever works best for your specific platform. Be
attentive not to use any PII in the transaction name. One recommendation is 

> `screen name + view identifier || accessibility identifier`. On Android, this would map to

Reviewer 2: `accessibility identifier` may have PII because of screen readers, do we guard this against `sendDefaultPii`?

Reviewer 2: For context, accessibility id was a problem for https://docs.sentry.io/platforms/react-native/touchevents/ and we needed to provide a way out, so it can't be used by default.

Reviewer: we only use the view identifier

Reviewer 2: @philipphofmann so I'd rather remove the `accessibility identifier` recommendation or guard it against `sendDefaultPii ` in the spec.
Also, you don't always have an identifier, in this case, we should write in the spec that you don't start a transaction at all.

Reviewer: yep, I agree we should mention that - we don't start any transactions if we're unable to find the view id. Can be improved, but usually the views that trigger action always have some ids on them to attach listeners to them.

Reviewer 2: @romtsn depends on the platform, it's not that common having an `id` (key) in Flutter unless people use it for UI tests or something because you set the listeners directly in the declaration.

Author: @marandaneto, on Cocoa, we use the method the UI element calls. I added this now. I also mention that `Choose whatever works best for your specific platform.`. Do you think that's fine or should I add more information, @marandaneto?

Reviewer 2: @philipphofmann https://github.com/getsentry/sentry-cocoa/blob/efeb69e9c50860c600d68640b1f465cc83a1a7f7/Sources/Sentry/SentryUIEventTracker.m#L130 and https://github.com/getsentry/sentry-cocoa/blob/d9757459a4a288ad57b2a8348def9eb839cbe72b/Sources/Sentry/SentryBreadcrumbTracker.m#L219
This can be PII and isn't guarded with `defaultPii`

Reviewer 2: > @marandaneto, on Cocoa, we use the method the UI element calls. I added this now. I also mention that `Choose whatever works best for your specific platform.`. Do you think that's fine or should I add more information, @marandaneto?

Yes, all good for the spec, but I guess we should raise an issue to guard the linked code against `sendDefaultPii` or remove it altogether, what do you think?

Author: I will raise issues, yes. 🥇 for finding them.

Author: https://github.com/getsentry/sentry-cocoa/issues/2406

    And the transaction already  is finished
    When the user triggers the same UI event
    Then the SDK starts a new UI event transaction

# The SDK adds auto-generated spans to the transaction bound to the scope.
# UI event transactions need auto-generated spans not to be empty. Therefore,
# the SDK can't start a UI event transaction if there is already a
# transaction bound to the scope by the user.
Scenario: Manually created transaction bound to the scope
    Given an ongoing manually created transaction by the user bound to the
        scope
    When the user triggers the a UI event
    Then the SDK doesn't start a UI event transaction
```

### Screen load transactions

This section deals specifically with auto-generated transactions for loading screens. If your SDK has
other types of auto-generated transactions, please update the specification here.

```Gherkin
Scenario: Ongoing UI event transaction
    Given an ongoing UI event transaction
    When the SDK creates a new screen load transaction
    Then the SDK finishes the ongoing UI event transaction
    And removes it from the scope
    And sets the status to canceled
    And waits for its children to finish

> Scenario: Ongoing screen load transaction

Reviewer: ```suggestion
Scenario: Ongoing screen load/navigation transaction
```

    Given an ongoing screen load transaction

>     When the SDK starts a new UI event transaction

Reviewer: @philipphofmann I guess we should also change it to `does not start a new UI event transaction` similar to Manually create transactions

auto-generated spans not to be empty. Therefore, the SDK shouldn't start a UI event transaction if there
is already an ongoing screen load/navigation transaction or a transaction bound to the scope by the user.

```Gherkin
Scenario: Same UI element with different event
    Given an ongoing UI event transaction
    When the user triggers the same UI element with a different event
    Or the user triggers a different UI element
    Then the SDK finishes the ongoing transaction
    And sets the status to OK
    And waits for the children to finish
    And cancels the idle timeout
    And removes the ongoing transaction from the scope
    And starts a new transaction
    And puts the new transaction on the scope

Scenario: UI event triggered but transaction ended
    Given an auto-generated transaction from any UI event
    And the transaction is already finished
    When the user triggers the same UI event
    Then the SDK starts a new UI event transaction

Scenario: Manually created transaction bound to the scope
    Given an ongoing manually created transaction by the user bound to the
        scope

>     When the user triggers the a UI event

Reviewer: ```suggestion
    When the user triggers a UI event
```

    Then the SDK doesn't start a UI event transaction
```

### Screen load/navigation transactions

This section deals specifically with auto-generated transactions for loading screens. If your SDK has
other types of auto-generated transactions, please update the specification here.

```Gherkin
Scenario: Ongoing UI event transaction
    Given an ongoing UI event transaction
    When the SDK creates a new screen load transaction
    Then the SDK finishes the ongoing UI event transaction
    And removes it from the scope
    And sets the status to canceled
    And waits for its children to finish

Scenario: Ongoing screen load/navigation transaction

>     Given an ongoing screen load/navigation transaction
>     When the SDK starts a new UI event transaction
>     Then the SDK doesn't start a UI event transaction

Reviewer: ```suggestion
    Given an ongoing screen load/navigation transaction
    When the user triggers the a UI event
    Then the SDK doesn't start a UI event transaction
```

# Specification

Users can change the `idleTimeout` via the SDK config options. The default is 3.0 seconds. 

The specification is written in the [Gherkin syntax](https://cucumber.io/docs/gherkin/reference/).

**1. Starting UI Event transactions** 
```
Given an instrumentable UI event
Then the SDK starts a UI event transaction 
And starts the idle timeout with idleTimeout of the options
```

**5. Wait for children when starting span**
```
Given a UI event transaction
When the SDK starts a child span
Then the SDK cancels the idle timeout
And waits for the child to finish
```

**4. Start timeout when the last span finishes**
```
Given a UI event transaction
And the transaction has one or multiple running child spans
And the transaction is waiting for its children to finish
When the SDK finishes the last child span
Then the SDK starts the idle timeout
```

**3. Don't reset timeout when the second last span finishes**
```
Given a UI event transaction
And the transaction has two running child spans
When the SDK finishes the first child span
Then the SDK doesn't reset the idle timeout

> And waits for the children to finish

Reviewer 2: H: This is different than Android and Flutter, actually, when a new span is created, the timer is reset only.
@romtsn to confirm.

Author: Doesn't Android also use `waitForChildren` @romtsn ?

Reviewer 2: There's `waitForChildren`, but the approach here is different if I understood it correctly, it does reset the timer every time that a new child is created and so on, at the end, the result is the same, I guess.

Reviewer: I guess I'm confused a bit by the word `reset` - does it mean `cancel` or does it mean `reschedule`? If we mean `reschedule` here, I think we can remove this scenario, because it should be covered by the one above ("when the _last_ span finishes")

That's what we do on Android (if all children are finished - we reschedule the finish timer):
https://github.com/getsentry/sentry-java/blob/84f0ef1abb55bccd90315665ff4e307c5f71886c/sentry/src/main/java/io/sentry/SentryTracer.java#L234-L245

Author: It's `reschedule`. 

>  I think we can remove this scenario, because it should be covered by the one above

Yes, it's covered. I removed it.

```

**2. Discard UI event transactions without child spans**
```
Given a UI event transaction
And the transaction has no child spans
When the idleTimeout times out
Then the SDK discards the transaction 
```

**3. Set time to last finished child span**
```
Given a UI event transaction
And the transaction has one finished child span
And the transaction has one running child span
When the running child span finishes
Then the SDK finishes the transaction
And trims the end time of the transaction to the one of the last finished child span

> ```

Reviewer 2: Android and Flutter have an option for that called `trimEnd` and it is set by default for idle transactions.

Author: Thanks for the info, but I guess it's OK to keep this? If you have `trimEnd` you are safe.

Reviewer 2: Maybe mention that some SDKs have the `trimEnd`, but you can also skip it and use the timestamp of the last finished child span as well.

Author: `trimEnd` what happens on Android if you disable it for that scenario, @romtsn?

Reviewer: It won't set the transaction end time to the latest child end time, so it will be set to `latest child end timestamp + idleTimeout`

Reviewer: but worth noting, it's not configurable from within SentryOptions, so it's internal on Android and is more of a convenience flag for us if we decide not to trim

Reviewer 2: Yeah, tbh `trimEnd` might be just overengineering, I guess we can skip this.

**6. Same UI element with same event**
```
Given an ongoing UI event transaction
When the user triggers the same UI event with the same type for the same UI element
Then the SDK resets the timeout
And doesn't create a new transaction
```

**10. Same UI element with different event**
If your SDK binds transactions to the scope see 11.
```
Given an ongoing UI event transaction
When the user triggers the same UI element with a different event
Or the user triggers a different UI element
Then the SDK finishes the ongoing transaction

> And sets the status to OK

Reviewer 2: I know this is already implemented, but it seems like the status could be `canceled` or something?
Also, the status is `ok` only if the user has not set a status manually.

Author: On Cocoa, we have `ok`, but we can change it.
```Swift
func test_DifferentUIElement_DifferentAction_FinishesTransaction() {
    let view1 = fixture.button
    callExecuteAction(action: "otherAction", target: fixture.target, sender: view1, event: TestUIEvent())
    
    let firstTransaction = try! XCTUnwrap(SentrySDK.span as? SentryTracer)
    
    let view2 = UIButton()
    callExecuteAction(action: action, target: fixture.target, sender: view2, event: TestUIEvent())
    
    assertFinishesTransaction(firstTransaction, operationClick)
}

private func assertFinishesTransaction(_ transaction: SentryTracer, _ operation: String) {
    XCTAssertTrue(transaction.isFinished)
    XCTAssertEqual(.ok, transaction.context.status)
    assertTransaction(name: "SentryTests.FirstViewController.\(expectedAction)", operation: operation)
    
    let transactions = getInternalTransactions()
    XCTAssertEqual(1, transactions.count)
}
```

https://github.com/getsentry/sentry-cocoa/blob/944bebeafd548a2cb391f8e6b2047ed37eec3755/Tests/SentryTests/Integrations/UIEvents/SentryUIEventTrackerTests.swift#L186-L196

https://github.com/getsentry/sentry-cocoa/blob/944bebeafd548a2cb391f8e6b2047ed37eec3755/Tests/SentryTests/Integrations/UIEvents/SentryUIEventTrackerTests.swift#L274

Reviewer 2: On Android, if I understood it correctly, it forces to `ok`.
@romtsn to confirm.

Reviewer: yes we use `ok`. I'm not sure about `canceled` because it doesn't sound like the user is cancelling/aborting their action - e.g. if they scrolled through twitter feed and found some interesting tweet and clicked to expand the thread - is it really cancelling the scrolling gesture, or just finishing the gesture logically, because there was an interesting thing requiring another gesture type?

`canceled` would make sense to me, if e.g. user clicked a button which triggered some network request, and this request was very slow to display the content, and the user pressed back/home button because they don't wanna wait anymore. Then it's for sure `canceled`, but we don't track system buttons yet.

Reviewer 2: If the agreement was `ok`, let's keep it as it is, but Android should only set a status if there's no status, right now its overwriting the user's status in case it's set, @romtsn mind filling a bug for this?

Reviewer 2: Btw just a small info, status `canceled` does not degrade the failure rate of transactions.
https://docs.sentry.io/product/performance/metrics/#failure-rate

Reviewer: https://github.com/getsentry/sentry-java/issues/2378 filed
